File: subsystem/statTools/cleanrrds.pl
#!/usr/bin/env perl
#
# $Id$
#
# Copyright 2003, 2004 Norwegian University of Science and Technology
#
# This file is part of Network Administration Visualized (NAV)
#
# NAV is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# NAV is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with NAV; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#
# Authors: John Magne Bredal <bredal@itea.ntnu.no>
#
# The purpose of this script is to search the cricket-data
# directory for rrd-files that are not in use anymore, and
# let the user delete these files if he wants.
#
# To accomplish this, we use find to locate the files, and
# then either lists them or deletes them in according to
# the users wishes.
#
# cleanrrds -h for help.
#

use warnings;
use vars qw($opt_l $opt_d $opt_t $opt_f $opt_h  $opt_p $opt_r);
use Getopt::Std;
use Pg;
use strict;

use NAV;

my $dbh = &NAV::connection('statTools', 'manage');

my $deletelimit = 30;
my $usage = "
cleanrrds finds rrd-files that has not been written to for a specified time,
and lists them or deletes them according to your wishes.

usage: $0 [-hld] [-t days] [-f string] [-p path]
\th : this helpstring
\tl : flag to list files
\td : flag to delete files (CAUTION! This will actually DELETE the files.)
\tr : flag to reverse-search, do a lookup in the database and delete entry if no rrd-file is found
\tt days     : selects files that has not been written to in t days
\tf string   : selects files that match the string
\tp path     : sets the path where we start looking\n";

getopts('rhldt:f:p:');

# Prints the help-text.
if ($opt_h) {
    print $usage;
    exit(0);
}

my $path = $opt_p;
$path = "/home/navcron/cricket/cricket-data/" unless $path;


if ($opt_r) {
    &reverseSearch();
    exit(0);
}

# Checking if both -d and -l option is selected
if ($opt_d && $opt_l) {
    print "You have selected both list and delete. The files will only be listed.\n";
}

# Setting default value of days if not set.
my $days = $opt_t;
$days = 30 unless $days;

# Checking if day is formatted correctly.
unless ($days =~ m/^\d+$/) {
    print $usage;
    exit(0);
}

# The list contains the output from the find-program.
my @list;
unless ($opt_f) {
    @list = `find $path -type f -mtime +$days -printf '%p %t %s\n' | grep '\\.rrd'`;
} else {
    @list = `find $path -type f -mtime +$days -printf '%p %t %s\n' | grep '\\.rrd' | grep $opt_f`;
}

my $teller = 0; # The number of files found
my $existteller = 0; # The number of files found also in the database
my $totalsize = 0; # The total size of all files in bytes.

# The general idea is this:
# If the file has been changed in the specified timeinterval, leave it.
# Otherwise delete

foreach my $line (@list) {

    chomp $line;
    $teller++;
    
    my $exists = 0;

    # Selects filepath,filename, time and size from the output.
    $line =~ m/(^\/.*)\/(.*\.rrd)\s(.*\d{2}:\d{2}:\d{2}(\.\d{10})?\s\d{4})\s(\d+)/;
    chomp (my $filepath = $1);
    chomp (my $filename = $2);
    chomp (my $time = $3);
    chomp (my $size = $5);

    #printf ("%s,%s,%s,%s\n", $1, $2, $3, $4);

    # calculates total size of all files.
    $totalsize += $size;

    # Looks in the database for the file.
    my $query = "SELECT rrd_fileid FROM rrd_file WHERE path='$filepath' AND filename='$filename'";
    my $r = $dbh->exec($query);
    
    my ($rrd_fileid) = $r->fetchrow;
    if ($r->ntuples > 0) {
	printf "%s,%s eksisterer i db.\n",$filepath,$filename;
	$exists = 1;
	$existteller++;
    }

    my $totalpath = "$filepath/$filename";

    # Lists the files
    if ($opt_l) {
	printf "%-70s %10s\n", $totalpath,$time;
    } elsif ($opt_d) {
	# Deletes the files.
	if (-e "$totalpath") {
	    print "Deleting $totalpath\n";
	    `rm -f \"$totalpath\"`;
	    
	    # Finds metafile
	    $totalpath =~ s/\.rrd/\.meta/; # somefile.rrd -> somefile.meta
	    if (-e "$totalpath") {
		print "Deleting $totalpath\n";
		`rm -f \"$totalpath\"`;
	    }
	}
	# Deleting from the database
	my $deleteq = "DELETE FROM rrd_file WHERE rrd_fileid = $rrd_fileid";
	my $deleter = $dbh->exec($deleteq);
    }
}
    
my $mbytes = int ($totalsize / (1024 * 1024));
if ($opt_f) {
    print "\n$teller rrd-files found that has not been modified in $days days, and matches string \"$opt_f\". $existteller of these are also in the database.\n";
} else {
    print "\n$teller rrd-files found that has not been modified in $days days. $existteller of these are also in the database.\n";
}
print "Total size = $totalsize bytes (~$mbytes MB).\n";


sub reverseSearch() {
    my $q = "SELECT rrd_fileid, path, filename FROM rrd_file WHERE subsystem='cricket'";
    my $r = $dbh->exec($q);

    my $existcounter = 0;
    my $deletedcounter = 0;

    while (my ($rrd_fileid, $path, $filename) = $r->fetchrow) {
	my $totalpath = "$path/$filename";
	if (-e $totalpath) {
	    $existcounter++;
	    print "$totalpath exists.\n";
	} else {
	    $deletedcounter++;
	    print "Could not find $totalpath, deleting it from db.\n";
	    my $deleteq = "DELETE FROM rrd_file WHERE rrd_fileid=$rrd_fileid";
	    print "$deleteq\n";
	    my $deleter = $dbh->exec($deleteq);
	}
    }

    printf "%s tuples checked, %s deleted, kept %s\n", $existcounter + $deletedcounter, $deletedcounter, $existcounter;

}

File: subsystem/statTools/cleanrrds.py
#!/usr/bin/env python
"""
This script does two things:
- Finds all rrd-files on the filesystem that has not been updated in a 
  given time, and deletes them.
- Finds all rrd-files in the database that does not exist on the filesystem
  and deletes the tuples in the database. 
"""

import re
import time
import os
from os.path import *
from optparse import OptionParser

from nav.db import getConnection 

# Default threshold in days for deleting files based on modification time.
#MODIFIEDTHRESHOLD = 365  
MODIFIEDTHRESHOLD = 0

def main(opts):

    if opts.db:
        clean_database()
    elif opts.path:
        clean_filesystem(opts.path, opts.delete)
        
def clean_database():
    """
    Search the rrd database for tuples that do not have a corresponding
    file. 
    """
    
    conn = getConnection("default")
    c = conn.cursor()
    
    query = """
    SELECT rrd_fileid, path, filename
    FROM rrd_file
    ORDER BY path, filename
    """
    c.execute(query)
    
    to_be_deleted = []
    for fileid, path, filename in c.fetchall():
        fullpath = join(path, filename)
        if not exists(fullpath):
            to_be_deleted.append((fileid, fullpath))
            
    if len(to_be_deleted) > 0:
        print "These files do not exist and will be removed from the database:"
        for id, fullpath in to_be_deleted:
            print fullpath
            if delete:
                c.execute("DELETE FROM rrd_file WHERE rrd_fileid=%s", (id,))
        conn.commit()
        print "%s tuples deleted from database." % len(to_be_deleted)
    else:
        print "All files existed, nothing done."
        
    return


def clean_filesystem(path, delete):
    """
    Finds all rrd files in path and deletes them if not modified the last 
    MODIFIEDTHRESHOLD days 
    """
    
    def clean_directory(delete, directory, files):
        """
        This method is called for every directory in path.
        """
        pattern = re.compile("\.rrd$")
        threshold = MODIFIEDTHRESHOLD * 24 * 60 * 60
        
        files_to_delete = []
        for file in files:
            filename = join(directory, file)
            if isdir(filename):
                continue
            
            if pattern.search(filename):
                # Check last modification time
                last_mod_time = getmtime(filename)
                time_since_mod = now - last_mod_time
                if time_since_mod > threshold:
                    print "%s has not been modified since %s." % (filename,
                                                    time.ctime(last_mod_time))
                    files_to_delete.append(filename)

        if len(files_to_delete) > 0:
            filesize = 0
            counter = 0
            for file in files_to_delete:
                try:
                    if delete:
                        os.remove(file)
                        filesize += getsize(file)
                        counter += 1
                        # Also remove .meta file if it exists
                        # Not crucial as it has no impact.
                        try:
                            metafile = re.sub('\.rrd$','.meta',file)
                            os.remove(metafile)
                        except:
                            pass
                except Exception, e:
                    print "Could not remove %s: %s" % (file, e)
                    continue
                
            if delete:
                print "Deleted %s files, freeing %s kbytes" % (counter,
                                                           int(filesize / 1024))
                    
    now = time.time()
    walk(path, clean_directory, delete)
    
    

if __name__ == '__main__':
    
    #===========================================================================
    # TODO: Make sure the options are the same as cleanrrds.pl?
    #===========================================================================
    
    usage = """usage: %%prog [options]
    Deletes rrd-files or database tuples based on input. Only lists files as 
    default, if you want to actually delete them use the --DELETE flag. The 
    default modification threshold for deleting a file is %s days. 
    """ % MODIFIEDTHRESHOLD
    
    parser = OptionParser(usage=usage)
    parser.add_option("-d", "--database", action="store_true", dest="db",
                      help="Clean database")
    parser.add_option("-f", "--filesystem", dest="path",
                      help="Clean filesystem given path to rrd-files.")
    parser.add_option("-t", "--threshold", dest="days",
                      help="Threshold in modification days for deleting files")
    parser.add_option("--DELETE", action="store_true", dest="delete",
                      default=False,
                  help="Set this flag to actually delete the files")
    
    opts, args = parser.parse_args()
    if not (opts.db or opts.path):
        parser.print_help()
    else:
        main(opts)
